2002-03-22  Jakub Jelinek  <jakub@redhat.com>

	PR c++/6037
	* decl.c (start_enum): Don't set TREE_ADDRESSABLE on TREE_LIST node.
	* g++.dg/other/enum4.C: New test.

--- gcc-2.95.2/gcc/cp/decl.c.c++-enum-fix	Fri Mar 22 16:29:15 2002
+++ gcc-2.95.2/gcc/cp/decl.c	Wed Mar 27 13:25:32 2002
@@ -13228,9 +13228,6 @@
       pushtag (name, enumtype, 0);
     }
 
-  if (current_class_type)
-    TREE_ADDRESSABLE (b->tags) = 1;
-
   GNU_xref_decl (current_function_decl, enumtype);
   return enumtype;
 }
--- gcc-2.95.2/gcc/testsuite/g++.old-deja/g++.other/enum4.C.c++-enum-fix	Wed Mar 27 13:25:32 2002
+++ gcc-2.95.2/gcc/testsuite/g++.old-deja/g++.other/enum4.C	Wed Mar 27 13:25:32 2002
@@ -0,0 +1,21 @@
+// Build don't link:
+
+/* PR c++/6037
+   This testcase ICEd because start_enum expected pushtag to insert
+   the tag always into current binding level.  */
+
+struct A
+{
+  ~A () { }
+};
+
+struct B
+{
+  void foo ()
+  {
+    switch (0) { default: ; }
+    A a;
+    enum C { };
+    (void) a;
+  }
+};
